rust/opendp/src/samplers.rs

```diff
@@ -502,17 +513,21 @@ impl<T: CastInternalReal + SampleRademacher + Zero> SampleLaplace for T {
             // adding a random sign to the exponential deviate does not induce gaps or stacks
             exponential * T::sample_standard_rademacher()?.into_internal()
         };
-
         rng.error?;
 
-        // initialize floats within mpfr/rug
-        let shift = shift.into_internal();
-        let scale = scale.into_internal();
-
-        // (shift / scale + noise) * scale. The noise itself is never scaled
-        let noised = shift.mul_add(&scale.clone().recip(), &laplace);
-        // postprocessing remains differentially private
-        Ok(Self::from_internal(noised * scale))
+        use rug::float::Round;
+        use rug::ops::{DivAssignRound, AddAssignRound};
+
+        // (shift / scale + noise) * scale.
+        let mut value = shift.into_internal();
+        // when scaling into the noise coordinate space, round down so that noise is overestimated
+        value.div_assign_round(&scale.clone().into_internal(), Round::Zero);
+        // the noise itself is never scaled. Round away from zero to offset the scaling bias
+        value.add_assign_round(
+            &laplace, if value.is_sign_positive() {Round::Up} else {Round::Down});
+        // postprocessing back to original coordinate space
+        //     remains differentially private via postprocessing
+        Ok(Self::from_internal(value) * scale)
```

Comment:
Should we factor this out, so that gaussian and Laplace can both use same logic?

Reply:
Factored out, added some more comments, and also censored the mpfr flags, FWIW.